src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/OpenDocumentGenerator.cs

```diff
@@ -0,0 +1,182 @@
+
+    private ProjectSnapshotManager? _projectManager;
+
+    public async Task DocumentOpenedOrChangedAsync(IDocumentSnapshot document, int version, CancellationToken cancellationToken)
+    {
+        // These flags exist to workaround things in VS Code, so bringing cohosting to VS Code without also fixing these flags, is very silly.
+        Debug.Assert(_languageServerFeatureOptions.IncludeProjectKeyInGeneratedFilePath);
+        Debug.Assert(!_languageServerFeatureOptions.UpdateBuffersForClosedDocuments);
+
+        // Actually do the generation
+        var generatedOutput = await document.GetGeneratedOutputAsync().ConfigureAwait(false);
+
+        // Now we have to update the LSP buffer etc.
+        // Fortunate this code will be removed in time
+        var hostDocumentUri = new Uri(document.FilePath);
+
+        _logger.LogDebug("[Cohost] Updating generated document buffers for {version} of {uri} in {projectKey}", version, hostDocumentUri, document.Project.Key);
+
+        if (_documentManager.TryGetDocument(hostDocumentUri, out var documentSnapshot))
+        {
+            // Html
+            var htmlVirtualDocumentSnapshot = TryGetHtmlSnapshot(documentSnapshot);
+
+            // CSharp
+            var csharpVirtualDocumentSnapshot = await TryGetCSharpSnapshotAsync(documentSnapshot, document.Project.Key, version, cancellationToken).ConfigureAwait(false);
+
+            // Buffer work has to be on the UI thread
+            await _joinableTaskContext.Factory.SwitchToMainThreadAsync(cancellationToken);
+
+            Debug.Assert(htmlVirtualDocumentSnapshot is not null && csharpVirtualDocumentSnapshot is not null ||
+                htmlVirtualDocumentSnapshot is null && csharpVirtualDocumentSnapshot is null, "Found a Html XOR a C# document. Expected both or neither.");
+
+            if (htmlVirtualDocumentSnapshot is not null)
+            {
+                _documentManager.UpdateVirtualDocument<HtmlVirtualDocument>(
+                    hostDocumentUri,
+                    [new VisualStudioTextChange(0, htmlVirtualDocumentSnapshot.Snapshot.Length, generatedOutput.GetHtmlSourceText().ToString())],
+                    version,
+                    state: null);
+            }
+
+            if (csharpVirtualDocumentSnapshot is not null)
+            {
+                _documentManager.UpdateVirtualDocument<CSharpVirtualDocument>(
+                    hostDocumentUri,
+                    csharpVirtualDocumentSnapshot.Uri,
+                    [new VisualStudioTextChange(0, csharpVirtualDocumentSnapshot.Snapshot.Length, generatedOutput.GetCSharpSourceText().ToString())],
+                    version,
+                    state: null);
+                return;
+            }
+        }
+    }
+
+    private async Task<CSharpVirtualDocumentSnapshot?> TryGetCSharpSnapshotAsync(LSPDocumentSnapshot documentSnapshot, ProjectKey projectKey, int version, CancellationToken cancellationToken)
```

Comment:
❓do we do accessibility ordering of any kind in razor? If so, then please put the private methods lower than public, otherwise carry on.

Reply:
For aesthetic reasons I actually with the `Initialize` was private, but if I make it an explicit implementation then it doesn't have any modifier, and that looks just as ugly, so I will move it.